apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/internal/DeferredJsonMerger.kt

```diff
@@ -0,0 +1,112 @@
+package com.apollographql.apollo3.internal
+
+import com.apollographql.apollo3.annotations.ApolloInternal
+import com.apollographql.apollo3.api.json.BufferedSourceJsonReader
+import com.apollographql.apollo3.api.json.readAny
+import okio.BufferedSource
+
+/**
+ * Utility class for merging GraphQL JSON payloads received in multiple chunks when using the `@defer` directive.
+ *
+ * Each call to [merge] will merge the given chunk into the [merged] Map, and will also update the [mergedFragmentLabels] Set with the
+ * value of its `label` field (if present).
+ *
+ * The fields in `data` are merged into the node found in [merged] at `path` (for the first call to [merge], the payload is
+ * copied to [merged] as-is).
+ *
+ * Errors are merged by adding them to [merged]'s `errors` list field.
+ *
+ * Extensions are merged by adding them into [merged]'s `extensions` field - if several payloads have the same key in `extensions`,
+ * their values are stored as a list.
```

Comment:
For errors/extensions, I _think_ I'd prefer only exposing the last value to users. If users need the list, they can always build it as payloads are received.

Reply:
Wouldn't it make the API a bit inconsistent if some parts are merged but not all parts? Do you see disadvantages of having the errors/extensions merged? In the case of extensions, I agree it's a bit awkward to have to create a list when the same key is used in several chunks...

Reply:
> Wouldn't it make the API a bit inconsistent if some parts are merged but not all parts?

The spec is inconsistent. It defines how to merge data but doesn't tell how to merge errors/extensions.

> Do you see disadvantages of having the errors/extensions merged?

Let's say you have 

```
payload1 - data1 - (error1, error2)

payload2 - data2
```

When receiving payload2, one has to make a diff with payload1 to see if there are any new errors. 

So if you want to handle errors per payload, it's more complicated. Maybe that's never going to happen but maybe it is, not sure.

We could even think of a scenario where the error gets "cleared":

```graphql 

query GetHero {
  hero (
    # get the details asap for optimistic rendering, might fail
    ...heroDetails
    # also get the details defered, will never fail
    ...heroDetails @defer
  }
}
```

Reply:
Merging errors and extensions:
  - Pros:
    - Consistency: everything is merged in the emitted ApolloResponse
    - Looking at the last response is as if there was no `@defer` at all (I was thinking of maybe making `ApolloRequest.execute` return the last element - perhaps it is not that useful and throwing is actually better)
    - Having centralized error handling may be easier?
  - Cons:
    - To know if an error is new (specific to a partial response), users would need to diff manually
    - No way to clear errors if some are present on the same fragment on payload n and not on payload n+1 (example above)
    - Merging extensions is awkward because they're maps

I have the impression that the cons outweigh the pros 😅 For me the big plus was the consistency one but maybe it's not a big deal after all, as long as the doc is clear.

So I'll remove it (we can always revisit this if we have second thoughts!).